File: frontend/src/lib/theme.test.ts
import { describe, expect, it } from "vitest";

import {
  applyTenantThemeVariables,
  fetchCurrentTenantTheme,
  normalizeLocalThemePreference,
  normalizeTenantTheme,
  sanitizeThemeVariables,
} from "./theme";

describe("theme contract", () => {
  it("normalizes invalid local theme preference to dark morfosis default", () => {
    expect(normalizeLocalThemePreference({ mode: "sepia", palette: "unknown" })).toEqual({
      mode: "dark",
      palette: "morfosis",
    });
  });

  it("keeps valid light mode preference and known palette", () => {
    expect(normalizeLocalThemePreference({ mode: "light", palette: "tokyo-night" })).toEqual({
      mode: "light",
      palette: "tokyo-night",
    });
  });

  it("accepts only allowlisted tenant css variables", () => {
    const safe = sanitizeThemeVariables({
      "--morfoschool-color-primary": "oklch(0.52 0.16 250)",
      "--morfoschool-color-accent": "#f5a524",
      background: "red",
      "--unknown": "blue",
    });

    expect(safe).toEqual({
      "--morfoschool-color-primary": "oklch(0.52 0.16 250)",
      "--morfoschool-color-accent": "#f5a524",
    });
  });

  it("rejects unsafe tenant variable values", () => {
    expect(
      sanitizeThemeVariables({
        "--morfoschool-color-primary": "red; background:url(javascript:alert(1))",
        "--morfoschool-color-accent": "oklch(0.68 0.18 70)",
      }),
    ).toEqual({ "--morfoschool-color-accent": "oklch(0.68 0.18 70)" });
  });

  it("normalizes backend tenant theme envelope", () => {
    const theme = normalizeTenantTheme({
      data: {
        theme: {
          tenantId: "tenant-1",
          preset: "morfoschools-default",
          primaryColor: "oklch(0.52 0.16 250)",
          accentColor: "oklch(0.68 0.18 70)",
          logoUrl: "",
          version: 4,
          cacheKey: "tenant_theme:tenant-1:v4",
          cssVariables: {
            "--morfoschool-color-primary": "oklch(0.52 0.16 250)",
            "--morfoschool-color-accent": "oklch(0.68 0.18 70)",
          },
        },
      },
    });

    expect(theme.tenantId).toBe("tenant-1");
    expect(theme.version).toBe(4);
    expect(theme.cssVariables["--morfoschool-color-primary"]).toBe("oklch(0.52 0.16 250)");
  });

  it("applies only sanitized tenant variables to a style target", () => {
    const writes: Array<[string, string]> = [];
    const target = { setProperty: (name: string, value: string) => writes.push([name, value]) };

    applyTenantThemeVariables(target, {
      "--morfoschool-color-primary": "#123456",
      background: "red",
    });

    expect(writes).toEqual([["--morfoschool-color-primary", "#123456"]]);
  });

  it("fetches current tenant theme with cookie credentials", async () => {
    const calls: Array<[string, RequestInit | undefined]> = [];
    const theme = await fetchCurrentTenantTheme({
      fetcher: async (input, init) => {
        calls.push([input, init]);
        return new Response(
          JSON.stringify({
            data: {
              theme: {
                tenantId: "tenant-1",
                preset: "morfoschools-default",
                primaryColor: "#123456",
                accentColor: "#f5a524",
                logoUrl: "",
                version: 2,
                cacheKey: "tenant_theme:tenant-1:v2",
                cssVariables: { "--morfoschool-color-primary": "#123456" },
              },
            },
          }),
          { status: 200, headers: { "Content-Type": "application/json" } },
        );
      },
    });

    expect(calls[0]).toEqual([
      "http://localhost:8080/api/v1/tenants/current/theme",
      { credentials: "include" },
    ]);
    expect(theme.version).toBe(2);
    expect(theme.cssVariables).toEqual({ "--morfoschool-color-primary": "#123456" });
  });
});

File: backend/internal/app/theme_test.go
package app

import "testing"

func TestValidateTenantThemeRejectsUnsafeColorInjection(t *testing.T) {
	theme := tenantTheme{Preset: "custom", PrimaryColor: "red; background:url(javascript:alert(1))", AccentColor: "oklch(0.68 0.18 70)", LogoURL: ""}

	if err := validateTenantTheme(theme); err == nil {
		t.Fatalf("expected unsafe primary color to be rejected")
	}
}

func TestValidateTenantThemeAcceptsSafeOklchAndHex(t *testing.T) {
	theme := tenantTheme{Preset: "morfoschools-default", PrimaryColor: "oklch(0.52 0.16 250)", AccentColor: "#f5a524", LogoURL: "https://cdn.example.test/logo.svg"}

	if err := validateTenantTheme(theme); err != nil {
		t.Fatalf("expected safe theme to pass validation: %v", err)
	}
}

func TestThemeCacheKeyIncludesTenantAndVersion(t *testing.T) {
	if got := themeCacheKey("tenant-1", 7); got != "tenant_theme:tenant-1:v7" {
		t.Fatalf("unexpected cache key: %s", got)
	}
}

func TestTenantThemeCSSVariablesUseSafeNames(t *testing.T) {
	theme := tenantTheme{TenantID: "tenant-1", Preset: "morfoschools-default", PrimaryColor: "oklch(0.52 0.16 250)", AccentColor: "#f5a524", LogoURL: "", Version: 3}
	vars := theme.CSSVariables()

	if vars["--morfoschool-color-primary"] != theme.PrimaryColor || vars["--morfoschool-color-accent"] != theme.AccentColor {
		t.Fatalf("unexpected css variable contract: %#v", vars)
	}
	if _, exists := vars["background"]; exists {
		t.Fatalf("css variable contract must not expose arbitrary property names: %#v", vars)
	}
}

func TestThemeCacheStoresByTenantVersion(t *testing.T) {
	cache := newMemoryThemeCache()
	theme := tenantTheme{TenantID: "tenant-1", Preset: "morfoschools-default", PrimaryColor: "oklch(0.52 0.16 250)", AccentColor: "#f5a524", Version: 9}

	if err := cache.Set(t.Context(), theme); err != nil {
		t.Fatalf("set cache: %v", err)
	}
	got, ok, err := cache.Get(t.Context(), theme.TenantID, theme.Version)
	if err != nil || !ok {
		t.Fatalf("expected cached theme, ok=%v err=%v", ok, err)
	}
	if got.CacheKey != "tenant_theme:tenant-1:v9" || got.PrimaryColor != theme.PrimaryColor {
		t.Fatalf("unexpected cached theme: %#v", got)
	}
}
